### PR TITLE
concat two map for edges

### PR DESCRIPTION
**Pull request by nobrakal**
The actual implementation of `edges` for `AdjacencyMap` used two successive `map`. Here is a reduced version :)

### CODE REVIEW

**Comment by snowleopard** [comment]
@nobrakal Thanks, but are you sure this is actually an issue in terms of performance? Could you check using your benchmarks? I'm pretty curious to see the results!

**Comment by nobrakal** [comment]
Just done it with a brand new `NFData` instance made from the #62  PR:
```Haskell
instance NFData a => NFData (AdjacencyMap a) where
  rnf (AM am) = rnf am
```

The `one map` version is quicker, but just a bit:
```
Doing:

----
make a Path from a list
make a Circuit from a list
make a Complete from a list

----

## make a Path from a list
### 1

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    276.4 ns | 0.999 |
| Two map ||    347.3 ns | 1.000 |
+---------++-------------+-------+

### 10

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    5.671 μs | 0.993 |
| Two map ||    5.978 μs | 0.997 |
+---------++-------------+-------+

### 100

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    77.86 μs | 0.999 |
| Two map ||    82.80 μs | 0.997 |
+---------++-------------+-------+

### 1000

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    1.260 ms | 0.997 |
| Two map ||    1.376 ms | 0.992 |
+---------++-------------+-------+

### 10000

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    21.38 ms | 0.955 |
| Two map ||    21.79 ms | 0.976 |
+---------++-------------+-------+


SUMMARY:

 * One map was the fastest 1 times
There was 4 ex-aequo


ABSTRACT:

 * One map was 1.10 times faster than Two map

## make a Circuit from a list
### 1

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    759.6 ns | 0.999 |
| Two map ||    850.1 ns | 1.000 |
+---------++-------------+-------+

### 10

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    5.985 μs | 0.999 |
| Two map ||    6.474 μs | 1.000 |
+---------++-------------+-------+

### 100

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    76.37 μs | 1.000 |
| Two map ||    80.87 μs | 1.000 |
+---------++-------------+-------+

### 1000

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    1.195 ms | 0.999 |
| Two map ||    1.253 ms | 0.999 |
+---------++-------------+-------+

### 10000

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    19.98 ms | 0.999 |
| Two map ||    20.15 ms | 0.999 |
+---------++-------------+-------+


SUMMARY:

 * One map was the fastest 1 times
There was 4 ex-aequo


ABSTRACT:

 * One map was 1.06 times faster than Two map

## make a Complete from a list
### 1

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    1.203 μs | 0.999 |
| Two map ||    1.390 μs | 1.000 |
+---------++-------------+-------+

### 10

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    55.33 μs | 1.000 |
| Two map ||    61.17 μs | 1.000 |
+---------++-------------+-------+

### 100

+---------++-------------+-------+
|         || Time (Mean) |    R² |
+=========++=============+=======+
| One map ||    11.30 ms | 0.999 |
| Two map ||    11.67 ms | 0.999 |
+---------++-------------+-------+


SUMMARY:

 * One map was the fastest 2 times
There was 1 ex-aequo


ABSTRACT:

 * One map was 1.10 times faster than Two map
```

**Comment by snowleopard** [comment]
@nobrakal Apologies for taking so long. Even though the improvement is not huge, I think it's worth making this improvement. Please address my comment and I'll merge.

File: src/Algebra/Graph/AdjacencyMap.hs [hunk 1]
```diff
@@ -191,9 +191,11 @@ connects = C.connects
 -- 'overlay' (fromAdjacencyList xs) (fromAdjacencyList ys) == fromAdjacencyList (xs ++ ys)
 -- @
 fromAdjacencyList :: Ord a => [(a, [a])] -> AdjacencyMap a
-fromAdjacencyList as = mkAM $ Map.unionWith Set.union vs es
+fromAdjacencyList = fromAdjacencyList' . map (fmap Set.fromList)
+
+fromAdjacencyList' :: Ord a => [(a, Set a)] -> AdjacencyMap a
```
[the remark this answers is not in the document]
Comment (nobrakal) [reply src/Algebra/Graph/AdjacencyMap.hs]: I added comments, and since it can be used on its own, I have exported it, and added a simple testsuite

**Comment by snowleopard** [comment]
@nobrakal Thanks, I might do some later polishing. 

The CI bots are failing though!

**Comment by nobrakal** [comment]
Wwwoooppps totally my bad, I forget a `$` in an equation, and pushed without checking .......

**Comment by snowleopard** [comment]
@nobrakal Merged, thanks!

**Comment by nobrakal** [comment]
Great, yeah this is better :)

**Pull request merged**